cmd/app.go

```diff
@@ -323,30 +325,49 @@ func appUpgrade(ctx *cli.Context) error {
 		return err
 	}
 
-	u, err := url.Parse(app.ExternalID)
-	if err != nil {
-		return err
+	au := &projectClient.AppUpgradeConfig{
+		Answers: answers,
 	}
 
-	q := u.Query()
+	if resolveTemplatePath(appVersionOrLocalTemplatePath) {
+		// if it is a path, upgrade install charts locally
+		localTemplatePath := appVersionOrLocalTemplatePath
+		_, files, err := walkTemplateDirectory(localTemplatePath)
+		if err != nil {
+			return err
+		}
 
-	q.Set("version", ctx.Args().Get(1))
+		au.Files = files
+	} else {
+		appVersion := appVersionOrLocalTemplatePath
 
-	filter := defaultListOpts(ctx)
-	filter.Filters["externalId"] = "catalog://?" + q.Encode()
+		// check external ID exists or not
+		externalID := app.ExternalID
+		if externalID == "" {
+			return fmt.Errorf("app %q is installed by local template folder", appName)
+		}
 
-	template, err := c.ManagementClient.TemplateVersion.List(filter)
-	if err != nil {
-		return err
-	}
+		u, err := url.Parse(externalID)
+		if err != nil {
+			return err
+		}
 
-	if len(template.Data) == 0 {
-		return fmt.Errorf("version %s not valid for app", ctx.Args().Get(1))
-	}
+		q := u.Query()
+		q.Set("version", appVersion)
 
-	au := &projectClient.AppUpgradeConfig{
-		Answers:    answers,
-		ExternalID: template.Data[0].ExternalID,
+		filter := defaultListOpts(ctx)
+		filter.Filters["externalId"] = u.String()
```

Comment:
The version in `u.String()` is the original one, not the version about to upgrade.

Reply:
Fixed